src/coreclr/jit/ee_il_dll.cpp

```diff
@@ -1628,7 +1628,19 @@ const WCHAR* Compiler::eeGetCPString(size_t strHandle)
         return (nullptr);
     }
 
-    CORINFO_String* asString = *((CORINFO_String**)strHandle);
+    CORINFO_String* asString = nullptr;
+    if (impGetStringClass() == *((CORINFO_CLASS_HANDLE*)strHandle))
+    {
+        // strHandle is a frozen string
+        // We assume strHandle is never an "interior" pointer in a frozen string
+        // (jit is not expected to perform such foldings)
+        asString = (CORINFO_String*)strHandle;
+    }
+    else
+    {
+        // strHandle is a pinned handle to a string object
+        asString = *((CORINFO_String**)strHandle);
```

Comment:
I'm a bit confused about this change, doesn't it mean that we are incorrectly marking something as a string _handle_? For me, the _handle_ part refers to the fact that there this is an address of a slot. So I would probably expect that raw object addresses should have some other handle type.

Reply:
@jakobbotsch after https://github.com/dotnet/runtime/pull/49576 string handles are either direct object references (in most cases) or pinned handles to string objects - in theory we could introduce _STR_HDL and _STR_PINNED_HDL to avoid doing this here but it's simply not worth it IMO

Reply:
What is the point of having `GTF_ICON_STR_HDL` if it does not tell us anything about the structure of the data that the pointer points to? Then it might just as well be marked as `GTF_ICON_CONST_PTR`, won't it?

Reply:
I would personally prefer to have `GTF_ICON_STR_OBJ` and `GTF_ICON_STR_HDL` in this case. Or `GTF_ICON_OBJ` that both strings and type objects (in #75573) could be marked as. The method table check you are doing in this code can be used to distinguish types once we have the latter handle type, if necessary.

Reply:
is it worth having a separate flag for debug/jitdisasm only method? I don't see we need it anywhere else

Reply:
`GTF_ICON_OBJ` - yeah, that's an option but I assume not as part of this specific PR

Reply:
Given that you are going to be adding another one in #75573 anyway that could reuse it I would say yes. We may want to add more later optimizations that makes use of the structure (for example, I guess we could optimize type checks on frozen objects eagerly?).

Reply:
Yeah, will do unification in https://github.com/dotnet/runtime/pull/75573